File: android/app/src/main/java/io/github/vezzulab/juke/data/Lyrics.kt
package io.github.vezzulab.juke.data

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.security.MessageDigest
import java.util.concurrent.TimeUnit

/** One line of a synced song: when it starts, what it says and, when the file has them, when each word starts. */
data class LyricLine(val start: Long, val text: String, val words: List<Pair<Long, String>>)

/** Lyrics: plain text or the synced ``.lrc`` format ("[01:23.45] words"), read the same way Juke for Linux reads them. */
object Lyrics {
    private val stamp = Regex("\\[(\\d{1,3}):(\\d{2})(?:[.:](\\d{1,3}))?]")
    private val wordStamp = Regex("<(\\d{1,3}):(\\d{2})(?:[.:](\\d{1,3}))?>")
    private val header = Regex("^\\[(?:ar|ti|al|au|by|offset|re|ve|length):.*]$", RegexOption.IGNORE_CASE)
    const val LEAD_IN_MS = 3000L
    const val GAP_MS = 8000L

    private fun ms(m: String, s: String, f: String): Long = m.toLong() * 60_000 + s.toLong() * 1000 + f.ifEmpty { "0" }.padEnd(3, '0').take(3).toLong()

    fun parse(text: String): List<LyricLine> {
        val lines = ArrayList<LyricLine>()
        for (raw in text.lines()) {
            val stamps = stamp.findAll(raw).toList()
            if (stamps.isEmpty()) continue
            val body = raw.substring(stamps.last().range.last + 1)
            val marks = wordStamp.findAll(body).toList()
            val words = marks.mapIndexedNotNull { i, mark ->
                val end = if (i + 1 < marks.size) marks[i + 1].range.first else body.length
                val word = body.substring(mark.range.last + 1, end).trim()
                if (word.isEmpty()) null else ms(mark.groupValues[1], mark.groupValues[2], mark.groupValues[3]) to word
            }
            val spoken = wordStamp.replace(body, "").trim()
            stamps.forEach { s -> lines += LyricLine(ms(s.groupValues[1], s.groupValues[2], s.groupValues[3]), spoken, words) }
        }
        return lines.sortedBy { it.start }
    }

    fun isSynced(text: String) = parse(text).size >= 2

    fun plain(text: String): String = text.lines().filterNot { header.matches(it.trim()) }
        .joinToString("\n") { if (stamp.containsMatchIn(it)) stamp.replace(it, "").trim() else it.trimEnd() }.trim('\n')

    fun lineAt(lines: List<LyricLine>, position: Long): Int = lines.indexOfLast { it.start <= position }

    /** How much of the line has been sung, 0..1: by word times when there are some, else evenly until the next line. */
    fun sweep(lines: List<LyricLine>, index: Int, position: Long): Float {
        if (index !in lines.indices) return 0f
        val line = lines[index]
        if (line.words.isNotEmpty() && line.text.isNotEmpty()) {
            val total = line.words.sumOf { it.second.length + 1 }
            var done = 0.0
            for (i in line.words.indices) {
                val (start, word) = line.words[i]
                val end = if (i + 1 < line.words.size) line.words[i + 1].first else lines.getOrNull(index + 1)?.start ?: (start + 600)
                if (position >= end) done += word.length + 1
                else { if (position > start) done += (word.length + 1) * (position - start).toDouble() / maxOf(1L, end - start); break }
            }
            return (done / total).toFloat().coerceIn(0f, 1f)
        }
        val following = lines.getOrNull(index + 1)?.start ?: (line.start + 5000)
        val span = maxOf(800L, minOf(following - line.start, 7000L)) * 0.92
        return ((position - line.start) / span).toFloat().coerceIn(0f, 1f)
    }

    /** 3, 2 or 1 while the next line is about to begin after a long enough silence (or the intro); else 0. */
    fun countdown(lines: List<LyricLine>, index: Int, position: Long): Int {
        val following = lines.getOrNull(index + 1)?.start ?: return 0
        val quietSince = if (index >= 0) lines[index].start else 0L
        if (index >= 0 && following - quietSince < GAP_MS) return 0
        val remaining = following - position
        return if (remaining in 1..LEAD_IN_MS) ((remaining + 999) / 1000).toInt() else 0
    }
}

/** What the person saved, one small file per song, kept apart from the music. */
class LyricsStore(context: Context) {
    private val dir = File(context.applicationContext.filesDir, "lyrics").apply { mkdirs() }

    private fun file(key: String) = File(dir, MessageDigest.getInstance("SHA-1").digest(key.toByteArray()).joinToString("") { "%02x".format(it) } + ".txt")

    fun get(key: String): String = runCatching { file(key).takeIf { it.exists() }?.readText().orEmpty() }.getOrDefault("")

    fun set(key: String, text: String) {
        runCatching { if (text.isBlank()) file(key).delete() else file(key).writeText(text.trim()) }
    }
}

/** LRCLIB (lrclib.net): a free, open lyrics service. It is asked only when the person asks, or has turned the automatic option on. */
object LrcLib {
    data class Found(val title: String, val artist: String, val album: String, val duration: Double, val plain: String, val synced: String) {
        val text get() = synced.ifBlank { plain }
        val isSynced get() = synced.isNotBlank()
    }

    private val client = OkHttpClient.Builder().connectTimeout(12, TimeUnit.SECONDS).readTimeout(15, TimeUnit.SECONDS).build()
    private const val BASE = "https://lrclib.net/api"
    private const val AGENT = "Juke-Android/0.3.3 (https://github.com/vezzulab/juke)"

    private fun found(o: JSONObject): Found? {
        if (o.optBoolean("instrumental")) return null
        val plain = o.optString("plainLyrics").takeIf { it != "null" }.orEmpty()
        val synced = o.optString("syncedLyrics").takeIf { it != "null" }.orEmpty()
        if (plain.isBlank() && synced.isBlank()) return null
        return Found(o.optString("trackName"), o.optString("artistName"), o.optString("albumName"), o.optDouble("duration", 0.0), plain, synced)
    }

    private suspend fun call(path: String, params: Map<String, String>): String? = withContext(Dispatchers.IO) {
        val url = "$BASE/$path".toHttpUrl().newBuilder().apply { params.forEach { (k, v) -> if (v.isNotBlank()) addQueryParameter(k, v) } }.build()
        client.newCall(Request.Builder().url(url).header("User-Agent", AGENT).build()).execute().use { r ->
            if (r.code == 404) null else if (!r.isSuccessful) throw java.io.IOException("HTTP ${r.code}") else r.body?.string()
        }
    }

    suspend fun search(title: String, artist: String): List<Found> {
        val body = call("search", mapOf("track_name" to title, "artist_name" to artist)) ?: return emptyList()
        val array = JSONArray(body)
        return (0 until array.length()).mapNotNull { found(array.getJSONObject(it)) }.take(25)
    }

    suspend fun exact(title: String, artist: String, album: String, durationSec: Int): Found? {
        val body = call("get", mapOf("track_name" to title, "artist_name" to artist, "album_name" to album, "duration" to durationSec.takeIf { it > 0 }?.toString().orEmpty())) ?: return null
        return found(JSONObject(body))
    }
}

File: juke/lyrics.py
"""Song lyrics: where they come from, the synced (.lrc) format and the LRCLIB search.

A song's lyrics are looked for in this order: what the person saved in Juke, a ``.lrc`` file beside the song,
then the lyrics tag inside the file. Searching LRCLIB (a free, open lyrics service that needs no account) only
happens when the person asks, or has turned on the automatic option in Settings; it sends the title, artist,
album and length of the song, nothing else.
"""

from __future__ import annotations

import bisect
import os
import re
from dataclasses import dataclass
from pathlib import Path

from . import __version__

LRCLIB = os.environ.get("JUKE_LYRICS_URL") or "https://lrclib.net/api"
_STAMP = re.compile(r"\[(\d{1,3}):(\d{2})(?:[.:](\d{1,3}))?\]")
_WORD_STAMP = re.compile(r"<(\d{1,3}):(\d{2})(?:[.:](\d{1,3}))?>")      # enhanced LRC: <mm:ss.xx>word
_TAGLINE = re.compile(r"^\[(?:ar|ti|al|au|by|offset|re|ve|length):.*\]$", re.IGNORECASE)


@dataclass(slots=True)
class Found:
    """One LRCLIB result."""
    title: str
    artist: str
    album: str
    duration: float
    plain: str
    synced: str

    @property
    def text(self) -> str:
        return self.synced or self.plain

    @property
    def is_synced(self) -> bool:
        return bool(self.synced)


def parse_lrc(text: str) -> list[tuple[int, str]]:
    """``[mm:ss.xx] words`` lines as (milliseconds, words), in time order. Empty when the text has no time stamps."""
    lines: list[tuple[int, str]] = []
    for raw in text.splitlines():
        stamps = list(_STAMP.finditer(raw))
        if not stamps:
            continue
        words = _WORD_STAMP.sub("", raw[stamps[-1].end():]).strip()
        for stamp in stamps:
            fraction = (stamp.group(3) or "0").ljust(3, "0")[:3]
            lines.append((int(stamp.group(1)) * 60_000 + int(stamp.group(2)) * 1000 + int(fraction), words))
    lines.sort(key=lambda item: item[0])
    return lines


@dataclass(slots=True)
class Line:
    """One line of a synced song: when it starts, what it says and, when the file has them, when each word starts."""
    start: int
    text: str
    words: list[tuple[int, str]]


def _stamp_ms(minutes: str, seconds: str, fraction: str | None) -> int:
    return int(minutes) * 60_000 + int(seconds) * 1000 + int((fraction or "0").ljust(3, "0")[:3])


def parse_lines(text: str) -> list[Line]:
    """The lines of a synced song for karaoke, with per-word times where the file gives them."""
    lines: list[Line] = []
    for raw in text.splitlines():
        stamps = list(_STAMP.finditer(raw))
        if not stamps:
            continue
        body = raw[stamps[-1].end():]
        words: list[tuple[int, str]] = []
        marks = list(_WORD_STAMP.finditer(body))
        for i, mark in enumerate(marks):
            end = marks[i + 1].start() if i + 1 < len(marks) else len(body)
            word = body[mark.end():end].strip()
            if word:
                words.append((_stamp_ms(*mark.groups()), word))
        spoken = _WORD_STAMP.sub("", body).strip()
        for stamp in stamps:
            lines.append(Line(_stamp_ms(*stamp.groups()), spoken, words))
    lines.sort(key=lambda line: line.start)
    return lines


LEAD_IN_MS = 3000            # the countdown dots start this long before a line
GAP_MS = 8000                # a silence at least this long between two lines is a break worth counting down


def sweep(lines: list[Line], index: int, position_ms: int) -> float:
    """How much of line ``index`` has been sung, 0..1: by its word times when it has them, otherwise evenly across
    the time until the next line (never slower than about 7 seconds)."""
    if not 0 <= index < len(lines):
        return 0.0
    line = lines[index]
    if line.words and line.text:
        total = sum(len(w) + 1 for _t, w in line.words)
        done = 0.0
        for i, (start, word) in enumerate(line.words):
            end = line.words[i + 1][0] if i + 1 < len(line.words) else (lines[index + 1].start if index + 1 < len(lines) else start + 600)
            if position_ms >= end:
                done += len(word) + 1
            elif position_ms > start:
                done += (len(word) + 1) * (position_ms - start) / max(1, end - start)
                break
            else:
                break
        return max(0.0, min(1.0, done / total))
    following = lines[index + 1].start if index + 1 < len(lines) else line.start + 5000
    span = max(800, min(following - line.start, 7000)) * 0.92
    return max(0.0, min(1.0, (position_ms - line.start) / span))


def countdown(lines: list[Line], index: int, position_ms: int) -> int:
    """3, 2 or 1 while the next line is about to start after a long enough silence (or the song's intro); else 0."""
    if not lines:
        return 0
    following = lines[index + 1].start if index + 1 < len(lines) else None
    if following is None:
        return 0
    quiet_since = lines[index].start if index >= 0 else 0
    if following - quiet_since < GAP_MS and index >= 0:
        return 0
    remaining = following - position_ms
    if 0 < remaining <= LEAD_IN_MS:
        return -(-remaining // 1000)          # 3, 2, 1
    return 0


def is_synced(text: str) -> bool:
    return len(parse_lrc(text)) >= 2


def plain_text(text: str) -> str:
    """The words only: time stamps and ``[ar:...]`` header lines removed."""
    out = []
    for raw in text.splitlines():
        line = _STAMP.sub("", raw).strip() if _STAMP.search(raw) else raw.rstrip()
        if _TAGLINE.match(raw.strip()):
            continue
        out.append(line)
    return "\n".join(out).strip("\n")


def line_at(timeline: list[tuple[int, str]], position_ms: int) -> int:
    """Index of the line being sung at ``position_ms`` (-1 before the first one)."""
    return bisect.bisect_right([ms for ms, _ in timeline], position_ms) - 1


def read_sidecar(path: str) -> str:
    """The ``.lrc`` file that sits beside the song, if there is one."""
    try:
        candidate = Path(path).with_suffix(".lrc")
        if candidate.is_file() and candidate.stat().st_size < 1_000_000:
            return candidate.read_text(encoding="utf-8", errors="replace").strip()
    except OSError:
        pass
    return ""


def read_embedded(path: str) -> str:
    """Lyrics stored in the file's own tags (ID3 USLT/SYLT-less text, Vorbis/FLAC LYRICS, MP4 ©lyr)."""
    import mutagen

    try:
        audio = mutagen.File(path)
    except Exception:
        return ""
    tags = getattr(audio, "tags", None)
    if audio is None or tags is None:
        return ""
    try:
        getall = getattr(tags, "getall", None)
        if getall is not None:                                     # ID3
            for frame in getall("USLT"):
                if str(frame.text).strip():
                    return str(frame.text).strip()
        for key in ("\xa9lyr", "LYRICS", "UNSYNCEDLYRICS", "lyrics", "unsyncedlyrics"):
            if key in tags:
                value = tags[key]
                first = value[0] if isinstance(value, (list, tuple)) and value else value
                if first and str(first).strip():
                    return str(first).strip()
    except Exception:
        return ""
    return ""


def local_lyrics(path: str) -> str:
    """Lyrics that come with the song itself: the ``.lrc`` beside it, else the tag inside it."""
    return read_sidecar(path) or read_embedded(path)


def _found(item: dict) -> Found | None:
    if not isinstance(item, dict) or item.get("instrumental"):
        return None
    plain, synced = str(item.get("plainLyrics") or ""), str(item.get("syncedLyrics") or "")
    if not (plain or synced):
        return None
    return Found(title=str(item.get("trackName") or ""), artist=str(item.get("artistName") or ""),
                 album=str(item.get("albumName") or ""), duration=float(item.get("duration") or 0), plain=plain, synced=synced)


async def search(title: str, artist: str = "", *, transport=None, timeout: float = 12.0) -> list[Found]:
    """LRCLIB results for a title (and artist), best matches first."""
    import httpx

    headers = {"User-Agent": f"Juke/{__version__} (https://github.com/vezzulab/juke)"}
    params = {"track_name": title}
    if artist:
        params["artist_name"] = artist
    async with httpx.AsyncClient(headers=headers, timeout=timeout, transport=transport, follow_redirects=True) as client:
        response = await client.get(f"{LRCLIB}/search", params=params)
        if response.status_code == 404:
            return []
        response.raise_for_status()
        data = response.json()
    return [f for f in (_found(item) for item in (data if isinstance(data, list) else [])) if f is not None][:25]


async def get_exact(title: str, artist: str, album: str, duration: float, *, transport=None, timeout: float = 12.0) -> Found | None:
    """The one LRCLIB entry that matches this song, using its length to tell versions apart."""
    import httpx

    headers = {"User-Agent": f"Juke/{__version__} (https://github.com/vezzulab/juke)"}
    params = {"track_name": title, "artist_name": artist, "album_name": album}
    if duration:
        params["duration"] = str(int(round(duration)))
    async with httpx.AsyncClient(headers=headers, timeout=timeout, transport=transport, follow_redirects=True) as client:
        response = await client.get(f"{LRCLIB}/get", params=params)
        if response.status_code == 404:
            return None
        response.raise_for_status()
        return _found(response.json())
